packages/angular_devkit/build_angular/src/utils/local-module-roots-tsconfig.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { Path, dirname, getSystemPath, join, normalize, resolve } from '@angular-devkit/core';
+import type { ParsedConfiguration } from '@angular/compiler-cli';
+import { promises as fs } from 'fs';
+
+let packageRootLookupCache: Map<Path, boolean> | null = null;
+
+/**
+ * Determines local module roots based on the given tsconfig configuration. A module
+ * root is a directory containing locally-built packages.
+ *
+ * The module roots are determined by inspecting the `paths` mappings in the config. These
+ * roots can then be passed to Webpack's resolution options so that e.g. local packages can
+ * benefit from Webpack module resolution that understands NodeJS `package.json` exports
+ */
+export async function findLocalModuleRootsFromTsconfig(
+  tsConfig: ParsedConfiguration,
+): Promise<string[]> {
+  if (tsConfig.options.paths === undefined) {
+    return [];
+  }
+
+  const basePath = normalize(tsConfig.options.baseUrl ?? '');
+  const roots = new Set<string>();
+
+  for (const [key, mappings] of Object.entries(tsConfig.options.paths)) {
+    for (const mapping of mappings) {
+      let target = resolve(basePath, normalize(mapping));
+
+      // Omit the trailing segments when a wildcard is used.
+      if (key[key.length - 1] === '*') {
+        target = target.slice(0, target.lastIndexOf('*')) as Path;
+      }
+
+      const packageRoot = await findPackageRoot(normalize(target), basePath);
```

Comment:
I think the `baseUrl` is not the best "boundary" here, so I should likely remove that. Ideally limit it to the workspace path